src/interpolation/details/ArborX_InterpDetailsSymmetricPseudoInverseSVD.hpp

```diff
@@ -0,0 +1,273 @@
+    // | s(p, p) | s(p, q) |   | a | b |
+    // +---------+---------+ = +---+---+
+    // | s(q, p) | s(q, q) |   | b | c |
+    // +---------+---------+   +---+---+
+
+    // Lets compute x, y and theta such that
+    // +---+---+              +---+---+
+    // | a | b |              | x | 0 |
+    // +---+---+ = R(theta) * +---+---+ * R(theta)^T
+    // | b | c |              | 0 | y |
+    // +---+---+              +---+---+
+
+    value_t cos, sin, x, y;
+    if (a == c)
+    {
+      cos = Kokkos::sqrt(value_t(2)) / 2;
+      sin = Kokkos::sqrt(value_t(2)) / 2;
+      x = a + b;
+      y = a - b;
+    }
+    else
+    {
+      value_t const u = (2 * b) / (a - c);
+      value_t const v = 1 / Kokkos::sqrt(u * u + 1);
+      cos = Kokkos::sqrt((1 + v) / 2);
+      sin = Kokkos::copysign(Kokkos::sqrt((1 - v) / 2), u);
+      x = (a + c + (a - c) / v) / 2;
+      y = a + c - x;
+    }
+
+    // Now lets compute the following new values for U amd E.S
+    // M <- R'(theta)^T . S . R'(theta)
+    // U <- U . R'(theta)
+
+    // R'(theta)^T . S . R'(theta)
+    std::size_t i = 0;
+    for (; i < p; i++)
+    {
+      value_t const s_ip = sigma(i, p);
+      value_t const s_iq = sigma(i, q);
+      sigma(i, p) = cos * s_ip + sin * s_iq;
+      sigma(i, q) = -sin * s_ip + cos * s_iq;
+    }
+    sigma(p, p) = x;
+    i++;
+    for (; i < q; i++)
+    {
+      value_t const s_pi = sigma(p, i);
+      value_t const s_iq = sigma(i, q);
+      sigma(p, i) = cos * s_pi + sin * s_iq;
+      sigma(i, q) = -sin * s_pi + cos * s_iq;
+    }
+    sigma(q, q) = y;
+    i++;
+    for (; i < size; i++)
+    {
+      value_t const s_pi = sigma(p, i);
+      value_t const s_qi = sigma(q, i);
+      sigma(p, i) = cos * s_pi + sin * s_qi;
+      sigma(q, i) = -sin * s_pi + cos * s_qi;
+    }
+    sigma(p, q) = 0;
+
+    // U . R'(theta)
+    for (std::size_t i = 0; i < size; i++)
+    {
+      value_t const o_ip = ortho(i, p);
+      value_t const o_iq = ortho(i, q);
+      ortho(i, p) = cos * o_ip + sin * o_iq;
+      ortho(i, q) = -sin * o_ip + cos * o_iq;
+    }
+  }
+
+  // We compute the max to get a range of the invertible eigen values
+  value_t max_eigen = epsilon;
+  for (std::size_t i = 0; i < size; i++)
+    max_eigen = Kokkos::max(Kokkos::abs(sigma(i, i)), max_eigen);
+
+  // We inverse the diagonal of S, except if "0" is found
+  for (std::size_t i = 0; i < size; i++)
+    sigma(i, i) =
+        (Kokkos::abs(sigma(i, i)) < max_eigen * epsilon) ? 0 : 1 / sigma(i, i);
+
+  // Then we fill out IO as the pseudo inverse
+  for (std::size_t i = 0; i < size; i++)
+    for (std::size_t j = 0; j < size; j++)
+    {
+      value_t tmp = 0;
+      for (std::size_t k = 0; k < size; k++)
+        tmp += sigma(k, k) * ortho(i, k) * ortho(j, k);
+      inv_matrix(i, j) = tmp;
+    }
+}
+
+template <typename ExecutionSpace, typename InvMatrices>
+void symmetricPseudoInverseSVD(ExecutionSpace const &space,
+                               InvMatrices &inv_matrices)
```

Comment:
I think the name `InvMatrices` is misleading, given that it is both input and output.

Reply:
I have renamed the type `InOutMatrices`, as the inverse is already implied by the function name.